### DISCUSSION TITLE
Dynamically change the system tray icon

### DISCUSSION THREAD

**Discussion by louremipsum** [tauri-apps/tauri#6527] [Q&A] [answered]
I am making a messaging app and I wanted a functionality where if there is a notification, the system tray icon can dynamically change based on the number of notifications. 
I have several .ico files for different numbers of notifications and I want to update the system tray icon accordingly

I saw the docs and there was written that to change the system tray icon, we can use 
```rust
app.tray_handle().set_icon(tauri::Icon::Raw(include_bytes!("../path/to/myicon.ico").to_vec())).unwrap();
```
but the path to the icon should be a string literal and not a variable so I am confused about how to change the icon 
i am new to Tauri and Rust so i might be missing something i don't know.

i found this repo https://github.com/tauri-apps/tray-icon but I'm not sure how to implement this

**Comment by FabianLars** [comment, marked as answer]
> i found this repo https://github.com/tauri-apps/tray-icon but I'm not sure how to implement this

This repo is irrelevant for you, it will be an implementation detail of tauri itself (as-in tauri's systray implementation will use that crate).

> but the path to the icon should be a string literal and not a variable so I am confused about how to change the icon

The confusion here may come from the include_bytes!() macro. it will include the myicon.ico file as bytes inside the app binary. So instead of the whole macro you can just insert your own .ico file _as bytes_. If you have the icon on the user's file system you could use the [`Icon::File`](https://docs.rs/tauri/latest/tauri/enum.Icon.html) variant to make tauri load the icon itself.

**Reply by louremipsum** [reply]
![image](https://user-images.githubusercontent.com/72456774/226934926-834bdc7a-3194-4cce-87fd-a6c1ff0b9152.png)
basically, I have a folder of different icons which ill be replacing on the number of notifications there is.
I know this is most probably not a good way to do this but this is giving an error that 
![image](https://user-images.githubusercontent.com/72456774/226931929-79bff7f3-a5a5-4e8f-b9cb-5573ee41b555.png)
also in what way can I make that path relative?

**Reply by FabianLars** [reply]
So the first question is how you want to bundle these files with your app then. The 2 main approaches that come to mind are the include_bytes! macro from above, and tauri's resource feature: https://tauri.app/v1/guides/building/resources.

With that you could do basically what you're doing in your screenshot but getting the correct icon via `app_handle.path_resolver().resolve_resource("path_to_ico")` (and then pipe that into Icon::File)

**Reply by louremipsum** [reply]
but ```rust app_handle.path_resolver().resolve_resource("path_to_ico")``` returns ```enum `Option<PathBuf>``` but Icon::FIle expects ```struct `PathBuf` ```  so its giving type-mismatch
![image](https://user-images.githubusercontent.com/72456774/226943408-1b06b0fd-c630-4ffd-a1ea-bfebbdb6e514.png)

![image](https://user-images.githubusercontent.com/72456774/226943755-ceb3f864-ad78-42ba-9a4d-20ed993d294c.png)

this is what I have to do right?

**Reply by FabianLars** [reply]
About the Option: You have to unwrap it in some way. If you _sure_ the path is correct you can literally just call unwrap() on it, if not there are other ways like `if let` or `match` https://doc.rust-lang.org/std/option/enum.Option.html

> this is what I have to do right?

The `resources` config looks fine, if it doesn't work try `tray/**`. That said, the input of resolve_resource must match the path of the config, in this case it should be `"tray/win32/notification-plus.ico"` i think.

**Reply by louremipsum** [reply]
![image](https://user-images.githubusercontent.com/72456774/226950196-95ad9278-b863-4a6d-9c83-9a0555949535.png)
thank you so much it finally worked now 
thanks for the speedy replies

**Reply by louremipsum** [reply]
One last thing
The original icons are 128x128 pixels but in the system tray, they are appearing kind of blurry. Is there any way Tauri can handle the sizes of the icons so they appear sharp as I expect that 128*128 is enough for a sharp system tray icon

**Reply by FabianLars** [reply]
is this a problem in dev or also in after `build`?

There could be 2 problems:
- 128px is too large for system trays, if it's used it may be downscaled in a weird way
- The ico has multiple layers and only the first one is used here (16x16 probably). At least for the main icon this should behave differently for dev vs build (or maybe that was just the window icon i really can't remember 🤔)

But really, i'm not happy with how tray and window icons are handled right now either so hopefully we can improve that in the future.

**Reply by mrexox** [reply]
When you update the icon in Tauri V2 the icon_as_template is reset. And when you call set_icon_as_template later you'll notice a blink (when the setting gets changed). It would be good to chain the settings for icon and icon_as_template.

**Comment by nongyehong** [comment]
hello,What should I do if I want this icon to blink?

**Reply by Xiaobaishushu25** [reply]
maybe you can use two different icon  alternating display， i guess

**Reply by nongyehong** [reply]
There is no flicker if you alternate two same ICONS, I think you could alternate a transparent icon with a colored one, but it doesn't seem to work
